services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,528 @@
+     *         </ul>
+     */
+    @Override
+    public Future<OperationResult<Void>> updateTenant(final String tenantId, final Tenant tenantObj,
+            final Optional<String> resourceVersion, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(tenantObj);
+        Objects.requireNonNull(resourceVersion);
+        Objects.requireNonNull(span);
+
+        return MongoDbDeviceRegistryUtils.isAllowedToModify(config, tenantId)
+                .compose(ok -> processUpdateTenant(tenantId, tenantObj, resourceVersion, span))
+                .recover(error -> Future.succeededFuture(MongoDbDeviceRegistryUtils.mapErrorToResult(error, span)));
+    }
+
+    private Future<OperationResult<Void>> processUpdateTenant(final String tenantId, final Tenant newTenant,
+            final Optional<String> resourceVersion, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(newTenant);
+        Objects.requireNonNull(resourceVersion);
+        Objects.requireNonNull(span);
+
+        final JsonObject updateTenantQuery = resourceVersion
+                .map(version -> new MongoDbDocumentBuilder().withVersion(version))
+                .orElse(new MongoDbDocumentBuilder())
+                .withTenantId(tenantId)
+                .document();
+
+        final Promise<JsonObject> updateTenantPromise = Promise.promise();
+        final TenantDto newTenantDto = new TenantDto(tenantId, newTenant,
+                new Versioned<>(newTenant).getVersion());
+        mongoClient.findOneAndReplaceWithOptions(config.getCollectionName(), updateTenantQuery,
+                JsonObject.mapFrom(newTenantDto), new FindOptions(), new UpdateOptions().setReturningNewDocument(true),
+                updateTenantPromise);
+
+        return updateTenantPromise.future()
+                .compose(updateResult -> Optional.ofNullable(updateResult)
+                        .map(updated -> {
+                            span.log(String.format("successfully updated tenant [%s]", tenantId));
+                            return Future.<OperationResult<Void>> succeededFuture(OperationResult.ok(
+                                    HttpURLConnection.HTTP_NO_CONTENT,
+                                    null,
+                                    Optional.empty(),
+                                    Optional.of(updateResult.getString(MongoDbDeviceRegistryUtils.FIELD_VERSION))));
+                        })
+                        .orElse(MongoDbDeviceRegistryUtils.checkForVersionMismatchAndFail(
+                                tenantId, resourceVersion, findTenant(tenantId))))
+                .recover((error) -> {
+                    if (error instanceof MongoCommandException) {
+                        logDuplicationError(((MongoCommandException) error).getErrorMessage(), tenantId, span);
```

Comment:
As we discussed, there is no need of checking here, if the duplication key error is due to _tenantId_ or  _subjectDN_. In case of update, there the duplicate error can happen only due to _subjectDN_. Hence I would suggest to remove this part and use `MongoDbDeviceRegistryUtils.isDuplicateKeyError()` as you check in the `processCreateTenant(...)`

Reply:
done